src/transformStateWithClones.js

```diff
@@ -7,7 +7,38 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  let copyState = { ...state };
+  const allPreviousVersionsOfState = [];
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        copyState = {
+          ...copyState,
+          ...action.extraData,
+        };
+        break;
+
+      case 'removeProperties':
+        for (const trash of action.keysToRemove) {
+          delete copyState[trash];
+        }
+        break;
+
+      case 'clear':
+        for (const key1 in copyState) {
+          delete copyState[key1];
+        }
+        break;
+
+      default:
+        break;
```

Comment:
`[CODE STYLE]: switch/case should always have default case for error handling.` Checklist p3.
Use `throw new Error` to send errors and try to describe error, you can use `action.type` which caused the error